cdap-master/src/main/java/co/cask/cdap/data/tools/DatasetUpgrader.java

```diff
@@ -96,6 +100,31 @@ public void upgrade() throws Exception {
 
     // Upgrade the datasets instance meta table
     datasetInstanceMDSUpgrader.upgrade();
+
+    // upgrade all the filesets base paths
+    for (DatasetSpecification fileSetSpec : datasetInstanceMDSUpgrader.getFileSetsSpecs()) {
+      upgradeFileSet(fileSetSpec);
+    }
+
+  }
+
+  /**
+   * Upgrades the {@link FileSet} and also its embedded filesets if any by moving the base path under
+   * namespaced directory
+   *
+   * @param dsSpec the {@link DatasetSpecification} of the {@link FileSet} to be upgraded
+   * @throws IOException
+   */
+  private void upgradeFileSet(DatasetSpecification dsSpec) throws IOException {
+    for (DatasetSpecification embeddedDsSpec : dsSpec.getSpecifications().values()) {
+      upgradeFileSet(embeddedDsSpec);
```

Comment:
And you cannot assume that every embedded dataset is a FileSet. Does the recursive call fail if you call this on a dataset that is not a FileSet?